File: core-java/src/main/java/com/kakao/actionbase/core/java/edge/EdgeState.java
package com.kakao.actionbase.core.java.edge;

import com.kakao.actionbase.core.java.annotation.AllowNulls;
import com.kakao.actionbase.core.java.annotation.Nullable;
import com.kakao.actionbase.core.java.state.StateValue;
import com.kakao.actionbase.core.java.state.TransitableState;
import com.kakao.actionbase.core.java.types.StructType;

import java.util.Map;

import org.immutables.value.Value;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;

@Value.Immutable
@JsonSerialize(as = ImmutableEdgeState.class)
@JsonDeserialize(as = ImmutableEdgeState.class)
public interface EdgeState extends TransitableState<EdgeEvent, EdgeState>, EdgeModel {

  @Override
  boolean active();

  @Override
  @Value.Default
  default long version() {
    return Long.MIN_VALUE;
  }

  @Override
  Object source();

  @Override
  Object target();

  @Override
  @Nullable
  Long createdAt();

  @Override
  @Nullable
  Long deletedAt();

  @Override
  @AllowNulls
  Map<String, StateValue> properties();

  @JsonIgnore
  @Value.Auxiliary
  default EdgePayload toPayload() {
    return Edges.toPayload(this);
  }

  @Override
  default EdgeState transit(EdgeEvent event, StructType schema) {
    return Edges.transit(this, event, schema);
  }
}

File: core/src/main/kotlin/com/kakao/actionbase/core/edge/EdgeState.kt
package com.kakao.actionbase.core.edge

import com.kakao.actionbase.core.edge.record.EdgeStateRecord
import com.kakao.actionbase.core.state.AbstractSchema
import com.kakao.actionbase.core.state.State

data class EdgeState(
    val source: Any,
    val target: Any,
    val state: State,
) {
    // for storage
    fun toRecord(tableCode: Int): EdgeStateRecord =
        EdgeStateRecord(
            key =
                EdgeStateRecord.Key.of(
                    tableCode = tableCode,
                    source = source,
                    target = target,
                ),
            value =
                EdgeStateRecord.Value(
                    active = state.active,
                    version = state.version,
                    createdAt = state.createdAt,
                    deletedAt = state.deletedAt,
                    properties =
                        state.properties
                            .map { (name, value) ->
                                AbstractSchema.codeOf(name) to value
                            }.toMap(),
                ),
        )
}
